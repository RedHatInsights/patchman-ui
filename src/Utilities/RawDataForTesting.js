// raw data for testing purposes only

export const advisoryRows = [{
    attributes: {
        advisory_type: 1,
        applicable_systems: 1,
        description: 'The microcode_ctl ',
        public_date: '2020-06-24T17:22:25Z',
        synopsis: 'microcode_ctl',
        severity: 3
    },
    id: 'RHEA-2020:2743',
    type: 'advisory'
}];

export const systemAdvisoryRows = [{
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
}];

export const systemRows = [{
    attributes: {
        display_name: 'automation_host',
        last_evaluation: '2020-08-12T13:57:54.028883Z',
        last_upload: '2020-08-12T09:26:33.891907Z',
        packages_installed: 0,
        packages_updatable: 0,
        rhba_count: 213,
        rhea_count: 35,
        rhsa_count: 77,
        stale: false
    },
    id: '8ddb54f5-aeeb-49b1-8448-c29049d686c1',
    type: 'system'
}];

export const advisoryHeader = {
    attributes: {
        advisory_type: 3,
        description: 'Kernel-based Virtual.',
        public_date: '2020-06-30T13:38:53Z',
        modified_date: '2020-06-30T13:38:53Z',
        severity: 3,
        synopsis: 'Important'
    },
    id: 'RHSA-2020:2774',
    type: 'advisory'
};

export const advisoryDetailRows = {
    data: {
        attributes: {
            cves: ['CVE-2020-11080'],
            description: 'libnghttp2.',
            fixes: null,
            modified_date: '2020-06-25T16:31:29Z',
            packages: { libnghttp2: '1.33.0-3.el8_2.1.x86_64' },
            public_date: '2020-06-25T16:31:29Z',
            references: [],
            severity: 3,
            solution: 'For details',
            synopsis: 'Important: ',
            topic: 'An update .'
        }
    },
    id: 'RHSA-2020:2755',
    type: 'advisory'
};
export const systemPackages = [{
    description: 'This package contains testing data',
    evra: '2.2.*',
    name: 'acl',
    summary: 'Access control list utilities',
    updates: null
}];

export const entityDetail = {
    loaded: true,
    activeApps: [],
    entity: {
        account: '60'
    }
};
