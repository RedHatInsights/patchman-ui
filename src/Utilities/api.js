import { encodeApiParams } from './Helpers';

export function createApiCall(
    endpoint,
    method,
    parameters = undefined,
    data = undefined
) {
    if (parameters && method === 'get') {
        endpoint = endpoint.concat(encodeApiParams(parameters));
    }

    let result = window.insights.chrome.auth
    .getUser()
    .then(() =>
        fetch('/api/patch/v1' + endpoint, {
            method,
            credentials: 'include',
            body: JSON.stringify(data)
        })
    )
    .then(res => {
        if (!res.ok) {
            const contentType = res.headers.get('content-type');
            if (contentType.indexOf('json') !== -1) {
                throw res;
            } else {
                throw {
                    errors: [
                        { status: res.status, detail: res.statusText }
                    ]
                };
            }
        }

        return res.json();
    })
    .catch(caughtError => {
        const error = Promise.resolve(caughtError || {});
        const genericError = {
            title:
                    'There was an error getting data'
        };
        return error.then(error => {
            const [info] = error.errors;
            const result = error.errors && { ...genericError, detail: info.detail, status: info.status } || genericError;
            throw result ;
        });
    });
    return result;
}

export const fetchApplicableAdvisoriesApi = params => {
    return createApiCall('/advisories', 'get', params);
};

export const fetchApplicableSystemAdvisoriesApi = params => {
    let { id, ...allParams } = params;
    return createApiCall(`/systems/${id}/advisories`, 'get', allParams);
};

export const fetchSystems = params => {
    return createApiCall('/systems', 'get', params);
};

export const fetchAdvisoryDetailsApi = params => {
    return createApiCall(`/advisories/${params.advisoryName}`, 'get');
};

export const fetchApplicablePackagesApi = params => {
    let { id, ...allParams } = params;
    return createApiCall(`/systems/${id}/packages`, 'get', allParams);
};

export const fetchAffectedSystems = params => {
    const { id, ...args } = params;
    return createApiCall(`/advisories/${id}/systems`, 'get', args);
};

export const fetchPackagesList = params => {
    return createApiCall('/packages', 'get', params);
};

export const exportAdvisoriesCSV = params => {
    let endpoint = '/export/advisories';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'text/csv' })
    }).then(res => res.text());
};

export const exportAdvisoriesJSON = params => {
    let endpoint = '/export/advisories';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'application/json' })
    }).then(res => res.json());
};

export const exportSystemsCSV = params => {
    let endpoint = '/export/systems';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'text/csv' })
    }).then(res => res.text());
};

export const exportSystemsJSON = params => {
    let endpoint = '/export/systems';
    endpoint = endpoint.concat(encodeApiParams(params));
    return fetch('/api/patch/v1' + endpoint, {
        method: 'get',
        credentials: 'include',
        headers: new Headers({ accept: 'application/json' })
    }).then(res => res.json());
};
