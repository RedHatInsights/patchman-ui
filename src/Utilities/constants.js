import BugIcon from '@patternfly/react-icons/dist/js/icons/bug-icon';
import EnhancementIcon from '@patternfly/react-icons/dist/js/icons/enhancement-icon';
import SecurityIcon from '@patternfly/react-icons/dist/js/icons/security-icon';
import React from 'react';
import { subtractDate } from './Helpers';

export const STATUS_REJECTED = 'rejected';
export const STATUS_LOADING = 'loading';
export const STATUS_RESOLVED = 'resolved';

export const compoundSortValues = {
    operating_system: {
        asc: 'os_name,os_major,os_minor',
        desc: '-os_name,-os_major,-os_minor'
    },
    applicable_advisories: {
        asc: 'rhsa_count,rhba_count,rhea_count',
        desc: '-rhsa_count,-rhba_count,-rhea_count'
    }
};

// messy check because of frontend-components tests
export const ENABLE_PACKAGES = window?.insights?.chrome?.isBeta && insights?.chrome?.isBeta();
export const storeListDefaults = {
    rows: [],
    metadata: {
        limit: 25,
        offset: 0,
        total_items: 0
    },
    expandedRows: {},
    selectedRows: {},
    queryParams: {},
    error: {}
};

export const publicDateOptions = [
    { apiValue: '', label: 'All', value: 'all' },
    {
        apiValue: `gt:${subtractDate(7)}`,
        label: 'Last 7 days',
        value: 'last7'
    },
    {
        apiValue: `gt:${subtractDate(30)}`,
        label: 'Last 30 days',
        value: 'last30'
    },
    {
        apiValue: `gt:${subtractDate(90)}`,
        label: 'Last 90 days',
        value: 'last90'
    },
    {
        apiValue: `gt:${subtractDate(365)}`,
        label: 'Last year',
        value: 'lastYear'
    },
    {
        apiValue: `lt:${subtractDate(365)}`,
        label: 'More than 1 year ago',
        value: 'moreThanYear'
    }
];

export const advisorySeverities = [
    {
        value: 0,
        label: 'N/A',
        color: 'var(--pf-global--Color--200)'
    },
    {
        value: 1,
        label: 'Low',
        color: 'var(--pf-global--Color--200)',
        text:
            'This rating is given to all other issues that have a security impact. These are the types of vulnerabilities' +
            ' that are believed to require unlikely circumstances to be able to be exploited, or where a successful exploit' +
            ' would give minimal consequences.'
    },
    {
        value: 2,
        label: 'Moderate',
        color: 'var(--pf-global--warning-color--100)',
        text:
            'This rating is given to flaws that may be more difficult to exploit but could still lead to some' +
            ' compromise of the confidentiality, integrity, or availability of resources, under certain circumstances.' +
            ' These are the types of vulnerabilities that could have had a Critical impact or Important impact' +
            ' but are less easily exploited based on a technical evaluation of the flaw, or affect unlikely' +
            ' configurations.'
    },
    {
        value: 3,
        label: 'Important',
        color: '#ec7a08',
        text:
            'This rating is given to flaws that can easily compromise the confidentiality, integrity, or availability' +
            ' of resources. These are the types of vulnerabilities that allow local users to gain privileges, allow' +
            ' unauthenticated remote users to view resources that should otherwise be protected by authentication,' +
            ' allow authenticated remote users to execute arbitrary code, or allow remote users to cause a denial' +
            ' of service.'
    },
    {
        value: 4,
        label: 'Critical',
        color: 'var(--pf-global--danger-color--100)',
        text:
            'This rating is given to flaws that could be easily exploited by a remote unauthenticated attacker' +
            ' and lead to system compromise (arbitrary code execution) without requiring user interaction.' +
            ' These are the types of vulnerabilities that can be exploited by worms. Flaws that require an authenticated' +
            ' remote user, a local user, or an unlikely configuration are not classed as Critical impact.'
    }
];

export const advisoryTypes = [
    {
        value: 3,
        label: 'Security',
        icon: <SecurityIcon />
    },
    {
        value: 2,
        label: 'Bugfix',
        icon: <BugIcon />
    },
    {
        value: 1,
        label: 'Enhancement',
        icon: <EnhancementIcon />
    }
];

export const updatableTypes = [
    {
        value: '',
        label: 'All'
    },
    {
        value: true,
        label: 'Upgradable'
    }
];

export const packagesListUpdatableTypes = [
    {
        value: '',
        label: 'All'
    },
    {
        value: 'gt:0',
        label: 'Upgradable'
    }
];

export const filterCategories = {
    advisory_type: {
        label: 'Advisory type',
        values: advisoryTypes
    },
    public_date: {
        label: 'Public date',
        values: publicDateOptions
    },
    updatable: {
        label: 'Status',
        values: updatableTypes
    },
    systems_updatable: {
        label: 'Status',
        values: packagesListUpdatableTypes
    }
};

export const entityTypes = {
    advisories: 'advisories',
    packages: 'packages'
};

export const ReadOnlyNotification = {
    title: 'title',
    detail: 'message'
};

export const remediationIdentifiers = {
    package: 'patch-package',
    advisory: 'patch-advisory'
};
