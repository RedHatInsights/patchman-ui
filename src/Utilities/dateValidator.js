import { intl } from './IntlProvider';
import messages from '../Messages';

/*
validates date of a type YYYY-MM-DD with starting limit 1990-01-01.
regex is used to comply the Patternfly date format.
*/
const dateValidator = (dateStr) => {
    const regex = /^(\d{4})-(0[1-9]|1[012])-(0?[1-9]|[12][0-9]|3[01])$/;

    if (dateStr && typeof dateStr === 'string' && dateStr.match(regex) === null) {
        return intl.formatMessage(messages.labelsErrorInvalidDate);
    }

    const date = new Date(dateStr);
    const timestamp = date.getTime();

    //Month had to be set to equal 0 to get the first month of the year.
    const minDate = new Date(1990, 0, 1);

    if (typeof timestamp !== 'number' || Number.isNaN(timestamp)) {
        return intl.formatMessage(messages.labelsErrorInvalidDate);
    }

    if (date < minDate) {
        return intl.formatMessage(messages.labelsErrorDateLimit);
    }

    return;
};

export default dateValidator;
