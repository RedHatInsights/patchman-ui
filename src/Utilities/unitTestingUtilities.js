import { useSelector } from 'react-redux';

/* eslint-disable */
export const initMocks = () => {

    window.insights = {
        loadInventory: (args) => new Promise((resolve, reject) =>  {
            resolve(({
                inventoryConnector: (args) => {
                            const InventoryTable = ({ children }) => <div >A mock passed! {children} </div>;
                            return ({ InventoryTable });
                        },
                        mergeWithEntities: () => {}
            }));
        }),
        loadRemediations: () => new Promise((resolve, reject) =>  {
            resolve({ 
                openWizard: () => new Promise((resolve, reject) =>  { 
                    resolve(true)
                })
            });
        })
    };

};

export const mockStore = (initialState, mutatedState) => {
    const customMiddleWare = store => next => action => {
        useSelector.mockImplementation(callback => {
            return callback(mutatedState);
        });
        next(action);
    };

    const mockStore = configureStore([customMiddleWare]);

    return mockStore(initialState);
}
/* eslint-enable */
