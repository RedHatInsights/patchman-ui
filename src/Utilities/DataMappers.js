import { processDate } from '@redhat-cloud-services/frontend-components-utilities/files/cjs/helpers';
import { flatMap } from 'lodash';
import React from 'react';
import AdvisoryType from '../PresentationalComponents/AdvisoryType/AdvisoryType';
import { EmptyAdvisoryList, EmptyPackagesList } from '../PresentationalComponents/Snippets/EmptyStates';
import { entityTypes } from './constants';
import { createUpgradableColumn, handlePatchLink } from './Helpers';
import { DescriptionWithLink } from '../PresentationalComponents/Snippets/DescriptionWithLink';

export const createAdvisoriesRows = (rows, expandedRows, selectedRows) => {
    if (rows.length !== 0) {
        return flatMap(rows, (row, index) => {
            return [
                {
                    id: row.id,
                    isOpen: expandedRows[row.id] === true,
                    selected: selectedRows[row.id] !== undefined,
                    cells: [
                        { title: handlePatchLink(entityTypes.advisories, row.id) },
                        { title: processDate(row.attributes.public_date) },
                        {
                            title: (
                                <AdvisoryType
                                    type={row.attributes.advisory_type}
                                />
                            )
                        },
                        {
                            title: handlePatchLink(
                                entityTypes.advisories,
                                row.id,
                                row.attributes.applicable_systems
                            )
                        },
                        row.attributes.synopsis
                    ]
                },
                {
                    cells: [
                        {
                            title: <DescriptionWithLink row = {row} />
                        }
                    ],
                    parent: index * 2
                }
            ];
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 5 },
                        title: <EmptyAdvisoryList />
                    }
                ]
            }
        ];
    }
};

export const createSystemAdvisoriesRows = (
    rows,
    expandedRows,
    selectedRows
) => {
    if (rows.length !== 0) {
        return flatMap(rows, (row, index) => {
            return [
                {
                    id: row.id,
                    isOpen: expandedRows[row.id] === true,
                    selected: selectedRows[row.id] !== undefined,
                    cells: [
                        { title: handlePatchLink(entityTypes.advisories, row.id) },
                        { title: processDate(row.attributes.public_date) },
                        {
                            title: (
                                <AdvisoryType
                                    type={row.attributes.advisory_type}
                                />
                            )
                        },
                        row.attributes.synopsis
                    ]
                },
                {
                    cells: [
                        {
                            title: <DescriptionWithLink row = {row} />
                        }
                    ],
                    parent: index * 2
                }
            ];
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 6 },
                        title: <EmptyAdvisoryList />
                    }
                ]
            }
        ];
    }
};

export const createSystemsRows = (rows, selectedRows = {}) => {
    const data =
        rows &&
        rows.map(row => {
            return {
                id: row.id,
                key: Math.random().toString() + row.id,
                packages_installed: row.attributes.packages_installed,
                applicable_advisories: [
                    row.attributes.rhea_count || 0,
                    row.attributes.rhba_count || 0,
                    row.attributes.rhsa_count || 0
                ],
                selected: selectedRows[row.id] !== undefined
            };
        });
    return data || [];
};

export const createPackageSystemsRows = (rows, selectedRows = {}) => {
    const data =
        rows &&
        rows.map(row => {
            return {
                id: row.id,
                key: Math.random().toString() + row.id,
                installed_evra: row.installed_evra,
                available_evra: row.updatable && row.available_evra || row.installed_evra,
                upgradable: row.updatable,
                selected: selectedRows[row.id] !== undefined
            };
        });
    return data || [];
};

export const createSystemPackagesRows = (rows, selectedRows = {}) => {
    if (rows.length !== 0) {
        return rows.map(pkg => {
            const pkgUpdates = pkg.updates || [];
            const latestUpdate = pkgUpdates[pkgUpdates.length - 1];

            return {
                id: pkg.name,
                key: pkg.name,
                selected: selectedRows[pkg.name] !== undefined,
                disableCheckbox: !latestUpdate,
                cells: [
                    { title: handlePatchLink(entityTypes.packages, pkg.name) },
                    { title: pkg.evra },
                    { title: (latestUpdate && latestUpdate.evra) || pkg.evra },
                    { title: createUpgradableColumn(pkg.updatable) },
                    { title: pkg.summary }
                ]
            };
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 7 },
                        title: <EmptyPackagesList />
                    }
                ]
            }
        ];
    }
};

export const createPackagesRows = (rows) => {
    if (rows.length !== 0) {
        return rows.map(pkg => {
            return {
                id: pkg.name,
                key: pkg.name,
                cells: [
                    { title: handlePatchLink(entityTypes.packages, pkg.name) },
                    { title: pkg.systems_installed },
                    { title: pkg.systems_updatable },
                    { title: pkg.summary }
                ]
            };
        });
    } else {
        return [
            {
                heightAuto: true,
                cells: [
                    {
                        props: { colSpan: 7 },
                        title: <EmptyPackagesList />
                    }
                ]
            }
        ];
    }
};
