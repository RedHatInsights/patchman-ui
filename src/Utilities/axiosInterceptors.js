import {
    authInterceptor,
    responseDataInterceptor
} from '@redhat-cloud-services/frontend-components-utilities/files/interceptors';
import axios from 'axios';
import { ReadOnlyNotification } from './constants';
const axiosInstance = axios.create();

export async function  accessCheckInterceptor (config)  {

    const userPermissions = await insights.chrome.getUserPermissions();
    const inventoryPermissionList = ['inventory:*:*', 'inventory:*:read', 'inventory:hosts:read'];

    const hasInventoryAccess = userPermissions.some((access) => inventoryPermissionList.includes(access?.permission));

    if (!hasInventoryAccess) {
        throw new axios.Cancel('Operation canceled. User does not have access to Inventory');
    }

    return config;
}

export function errorInterceptor(err) {

    if (!axios.isCancel(err)) {

        const { response, isAxiosError } = { ...err };

        if (response && isAxiosError) {
            const { status, statusText, data } = response;

            if (!status) {
                throw err;
            } else {
                const genericError = {
                    title:
                            'There was an error getting data'
                };

                const result = { ...genericError, detail: data.error && data.error || statusText, status };

                throw result;
            }
        }

        throw err;
    }
}

export function readOnlyInterceptor(error) {
    if (error.response && error.response.status === 503) {
        const data = ReadOnlyNotification;
        throw data;
    }

    throw error;
}

axiosInstance.interceptors.request.use(accessCheckInterceptor);
axiosInstance.interceptors.request.use(authInterceptor);
axiosInstance.interceptors.response.use(responseDataInterceptor);
axiosInstance.interceptors.response.use(null, readOnlyInterceptor);
axiosInstance.interceptors.response.use(null, errorInterceptor);

export default axiosInstance;
