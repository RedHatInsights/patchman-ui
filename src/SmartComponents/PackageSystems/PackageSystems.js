import { Button, ToolbarGroup, ToolbarItem } from '@patternfly/react-core';
import { AnsibeTowerIcon } from '@patternfly/react-icons';
import { TableVariant } from '@patternfly/react-table';
import { Unavailable } from '@redhat-cloud-services/frontend-components';
import { InventoryTable } from '@redhat-cloud-services/frontend-components/components/cjs/Inventory';
import propTypes from 'prop-types';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import messages from '../../Messages';
import searchFilter from '../../PresentationalComponents/Filters/SearchFilter';
import statusFilter from '../../PresentationalComponents/Filters/StatusFilter';
import { getStore, register } from '../../store';
import { changePackageSystemsParams, clearPackageSystemsStore, fetchPackageSystemsAction } from '../../store/Actions/Actions';
import { packagesSystemsInventoryReducer } from '../../store/Reducers/InventoryEntitiesReducer';
import { fetchPackageSystems } from '../../Utilities/api';
import { STATUS_REJECTED, STATUS_RESOLVED, remediationIdentifiers } from '../../Utilities/constants';
import { createPackageSystemsRows } from '../../Utilities/DataMappers';
import { arrayFromObj, buildFilterChips, createSortBy, remediationProvider } from '../../Utilities/Helpers';
import {
    useDeepCompareEffect, useHandleRefresh, useOnSelect, usePagePerPage,
    useRemoveFilter, useSortColumn
} from '../../Utilities/Hooks';
import { intl } from '../../Utilities/IntlProvider';
import RemediationModal from '../Remediation/RemediationModal';
import { packageSystemsColumns } from '../Systems/SystemsListAssets';

const PackageSystems = ({ packageName }) => {
    const dispatch = useDispatch();
    const enableRemediation = false;
    const [
        RemediationModalCmp,
        setRemediationModalCmp
    ] = React.useState(() => () => null);
    const rawPackageSystems = useSelector(
        ({ PackageSystemsStore }) => PackageSystemsStore.rows
    );
    const status = useSelector(
        ({ PackageSystemsStore }) => PackageSystemsStore.status
    );
    const selectedRows = useSelector(
        ({ PackageSystemsStore }) => PackageSystemsStore.selectedRows
    );
    const hosts = React.useMemo(
        () => createPackageSystemsRows(rawPackageSystems, selectedRows),
        [rawPackageSystems]
    );
    const metadata = useSelector(
        ({ PackageSystemsStore }) => PackageSystemsStore.metadata
    );
    const queryParams = useSelector(
        ({ PackageSystemsStore }) => PackageSystemsStore.queryParams
    );

    const inventoryColumns = useSelector(
        ({ entities }) => entities && entities.columns
    );

    const handleRefresh = useHandleRefresh(metadata, apply);
    const { filter, search } = queryParams;

    React.useEffect(() => {
        return () => dispatch(clearPackageSystemsStore());
    }, []);

    useDeepCompareEffect(() => {
        dispatch(
            fetchPackageSystemsAction({ id: packageName, ...queryParams })
        );
    }, [queryParams]);

    const [page, perPage] = usePagePerPage(metadata.limit, metadata.offset);

    function apply(params) {
        dispatch(changePackageSystemsParams(params));
    }

    const removeFilter = useRemoveFilter(filter, apply);

    const filterConfig = {
        items: [
            searchFilter(apply, search, intl.formatMessage(messages.labelsFiltersSystemsSearch)),
            statusFilter(apply, filter)
        ]
    };

    const activeFiltersConfig = {
        filters: buildFilterChips(filter, search),
        onDelete: removeFilter
    };

    const showRemediationModal = data => {
        setRemediationModalCmp(() => () => <RemediationModal data={data} />);
    };

    const constructFilename = (system) => {
        return `${packageName}-${system.available_evra}`;
    };

    const fetchAllData = () =>
        fetchPackageSystems({ id: packageName, limit: 999999 });

    const selectRows = (toSelect) => {
        dispatch({ type: 'SELECT_ENTITY', payload: toSelect });
    };

    const onSelect = enableRemediation && useOnSelect(rawPackageSystems,  selectedRows,
        fetchAllData, selectRows, constructFilename);

    // This is used ONLY for sorting purposes
    const getMangledColumns = () => {
        let updated = inventoryColumns && inventoryColumns.filter(({ key }) => key === 'updated')[0];
        updated = { ...updated, key: 'last_upload' };
        return [...packageSystemsColumns, updated];
    };

    const onSort = useSortColumn(getMangledColumns(), apply, 0);
    const sortBy = React.useMemo(
        () => createSortBy(getMangledColumns(), metadata.sort, 0),
        [metadata.sort]
    );

    const selectedCount = selectedRows && arrayFromObj(selectedRows).length;

    return (
        <React.Fragment>
            {status === STATUS_REJECTED ? <Unavailable/> : (
                <InventoryTable
                    onLoad={({ mergeWithEntities }) => {
                        const store = getStore();
                        register({
                            ...mergeWithEntities(
                                packagesSystemsInventoryReducer(packageSystemsColumns, store.getState().PackageSystemsStore)
                            )
                        });

                    }}
                    items={hosts}
                    page={page}
                    total={metadata.total_items}
                    perPage={perPage}
                    onRefresh={handleRefresh}
                    isLoaded={status === STATUS_RESOLVED}
                    tableProps = {{ canSelectAll: false, onSort, sortBy, onSelect,
                        variant: TableVariant.compact, className: 'patchCompactInventory' }}
                    filterConfig={filterConfig}
                    activeFiltersConfig = {activeFiltersConfig}
                    bulkSelect={enableRemediation && onSelect && {
                        count: selectedCount,
                        items: [{
                            title: `Select none (0)`,
                            onClick: () => {
                                onSelect('none');
                            }
                        }, {
                            title: `Select page (${hosts.length})`,
                            onClick: () => {
                                onSelect('page');
                            }
                        },
                        {
                            title: `Select all (${metadata.total_items})`,
                            onClick: () => {
                                onSelect('all');
                            }
                        }],
                        onSelect: (value) => {
                            value ? onSelect('all') : onSelect('none');
                        },
                        toggleProps: {
                            'data-ouia-component-type': 'bulk-select-toggle-button'
                        },
                        checked: selectedCount === metadata.total_items ? true : selectedCount === 0 ? false : null
                    }}
                >
                    {enableRemediation &&
                    <ToolbarGroup>
                        <ToolbarItem>
                            <Button
                                className={'remediationButtonPatch'}
                                isDisabled={
                                    arrayFromObj(selectedRows).length === 0
                                }
                                onClick={() =>
                                    showRemediationModal(
                                        remediationProvider(
                                            packageName,
                                            Object.keys(selectedRows),
                                            remediationIdentifiers.package
                                        )
                                    )
                                }
                                ouiaId={'toolbar-remediation-button'}
                            >
                                <AnsibeTowerIcon/>&nbsp;{intl.formatMessage(messages.labelsRemediate)}
                            </Button>
                            <RemediationModalCmp />
                        </ToolbarItem>
                    </ToolbarGroup>
                    }
                </InventoryTable>
            )}
        </React.Fragment>
    );
};

PackageSystems.propTypes = {
    packageName: propTypes.string
};

export default PackageSystems;
