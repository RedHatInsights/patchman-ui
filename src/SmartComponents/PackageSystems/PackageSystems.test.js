/* eslint-disable no-unused-vars */
import { act } from 'react-dom/test-utils';
import { Provider, useSelector } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { exportPackageSystemsCSV, exportPackageSystemsJSON, fetchPackageSystems } from '../../Utilities/api';
import { systemRows } from '../../Utilities/RawDataForTesting';
import { initMocks } from '../../Utilities/unitTestingUtilities.js';
import PackageSystems from './PackageSystems';

/* eslint-disable */
initMocks()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: jest.fn().mockReturnValue({ pathname: ' testName' })
}));

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn()
}));

jest.mock('@redhat-cloud-services/frontend-components-utilities/helpers', () => ({
    ...jest.requireActual('@redhat-cloud-services/frontend-components-utilities/helpers'),
    downloadFile: jest.fn()
}));


jest.mock('@redhat-cloud-services/frontend-components/Inventory', () => ({
    ...jest.requireActual('@redhat-cloud-services/frontend-components/Inventory'),
    InventoryTable: jest.fn(() => <div className='testInventroyComponentChild'><div>This is child</div></div>)
}));

jest.mock('../../Utilities/api', () => ({
    ...jest.requireActual('../../Utilities/api'),
    exportPackageSystemsCSV: jest.fn(() => Promise.resolve({ success: true }).catch((err) => console.log(err))),
    exportPackageSystemsJSON: jest.fn(() => Promise.resolve({ success: true }).catch((err) => console.log(err))),
    fetchPackageSystems: jest.fn(() => Promise.resolve({
        data: [{
            attributes: {
                advisory_type: 2,
                description: "The tzdata penhancements.",
                public_date: "2020-10-19T15:02:38Z",
                synopsis: "tzdata enhancement update"
            },
            id: "RHBA-2020:4282",
            type: "advisory"
        }]
    }).catch((err) => console.log(err))),
}));

const mockState = {
    entities: {
        rows: systemRows,
        metadata: {
            limit: 25,
            offset: 0,
            total_items: 10
        },
        expandedRows: {},
        selectedRows: { 'RHSA-2020:2774': true },
        error: {},
        status: 'resolved',
    },
    PackageSystemsStore: {
        queryParams: {},
    }
};

const initStore = (state) => {
    const customMiddleWare = () => next => action => {
        useSelector.mockImplementation(callback => {
            return callback(mockState);
        });
        next(action);
    };
    const mockStore = configureStore([customMiddleWare]);
    return mockStore(mockState);
}

let wrapper;
let store = initStore(mockState);

beforeEach(() => {
    console.error = () => { };
    store.clearActions();
    useSelector.mockImplementation(callback => {
        return callback(mockState);
    });
    wrapper = mount(<Provider store={store}>
        <Router><PackageSystems packageName='testName'/></Router>
    </Provider>);
});

afterEach(() => {
    useSelector.mockClear();
});

describe('PackageSystems.js', () => {

    it('Should dispatch change package systems params  action once only', () => {
        const dispatchedActions = store.getActions();
        expect(dispatchedActions.filter(item => item.type === 'CHANGE_PACKAGE_SYSTEMS_PARAMS')).toHaveLength(1);
    });


    it('Should open remediation modal', () => {

        const { dedicatedAction } = wrapper.find('.testInventroyComponentChild').parent().props();
        act(() => dedicatedAction.props.onClick(null, null, {
            id: "patch-advisory:RHBA-2020:4282"
        }));
        expect(wrapper.update().find('RemediationModal')).toBeTruthy();
    });

    describe('test exports', () => {

        global.Headers = jest.fn();
        global.fetch = jest.fn(() => Promise.resolve({ success: true }).catch((err) => console.log(err)));

        it('Should download csv file', () => {
            const { exportConfig } = wrapper.find('.testInventroyComponentChild').parent().props();
            exportConfig.onSelect(null, 'csv');
            expect(exportPackageSystemsCSV).toHaveBeenCalledWith({}, 'testName');
        });

        it('Should download json file', () => {
            const { exportConfig } = wrapper.find('.testInventroyComponentChild').parent().props();
            exportConfig.onSelect(null, 'json');
            expect(exportPackageSystemsJSON).toHaveBeenCalledWith({}, 'testName');
        });
    });

    describe('test entity selecting', () => {
        it('Should unselect all', () => {
            const { bulkSelect } = wrapper.find('.testInventroyComponentChild').parent().props();

            bulkSelect.items[0].onClick();
            const dispatchedActions = store.getActions();

            expect(dispatchedActions[3].type).toEqual('SELECT_ENTITY');
            expect(bulkSelect.items[0].title).toEqual('Select none (0)');
        });

        it('Should select a page', async () => {

            const { bulkSelect } = wrapper.find('.testInventroyComponentChild').parent().props();

            bulkSelect.items[1].onClick();
            const dispatchedActions = store.getActions();
            expect(dispatchedActions[2].type).toEqual('SELECT_ENTITY');
            expect(bulkSelect.items[1].title).toEqual('Select page (1)');
        });

        it('Should select all', async () => {

            const { bulkSelect } = wrapper.find('.testInventroyComponentChild').parent().props();

            bulkSelect.items[2].onClick();
            expect(fetchPackageSystems).toHaveBeenCalled();
            expect(bulkSelect.items[2].title).toEqual('Select all (0)');
        });

    });

});
/* eslint-enable */
