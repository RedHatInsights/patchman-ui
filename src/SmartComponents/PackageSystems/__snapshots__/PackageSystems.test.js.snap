// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageSystems.js Should open remediation modal 1`] = `
<AsyncRemediationButton
  isDisabled={false}
  remediationProvider={[Function]}
/>
`;
